### PR TITLE
Separating MBConvBlock, FusedMBConvBlock and Refactoring EfficientNetV2

### PR DESCRIPTION
**Pull request by DavidLandup0**
# What does this PR do?

There's a design issue that's not aligned with KCV's principles. MBConv blocks, as well as well as Fused MBConv blocks are general-purpose blocks, introduced with MobileNets and then re-used in practically all mobile-oriented/efficient-desired architectures since. Currently, we only have them as part of EfficientNets, in which they're proprietary methods.

If we don't separate them, we'll accumulate tech debt that'll be only harder to recover from down the line.

Since MaxViT (currently WIP), MobileNets and various other future architectures will use them (and we'd want to expose these to the user as well as a common building block), I propose that we separate them into standalone layers, which is this PR.

I've refactored EfficientNetV2 to use layers instead of methods and they're identical in param counts, as validation that they behave the same. One issue to figure out is:

- Original effnets list each layer within each MBConvBlock
- New MBConvBlock layer encapsulates them within a single `MBConvBlock`

Because of this, the total number of layers differs (151 vs 26). We can:

1. Expose all layers within the `MBConvBlock` and `FusedMBConvBlock` and add them to the EffNet while it's being built (so as to be able to load weights as they are right now)
2. Keep layers encapsulated and port our own weights (or re-train effnets)

Here's the difference:

```
input_3 (InputLayer)        [(None, 224, 224, 3)]     0         
                                                                 
 rescaling_2 (Rescaling)     (None, 224, 224, 3)       0         
                                                                 
 stem_conv (Conv2D)          (None, 112, 112, 32)      864       
                                                                 
 stem_bn (BatchNormalization  (None, 112, 112, 32)     128       
 )                                                               
                                                                 
 stem_activation (Activation  (None, 112, 112, 32)     0         
 )                                                               
                                                                 
 block1a_ (FusedMBConvBlock)  (None, 112, 112, 16)     576       
                                                                 
 block2a_ (FusedMBConvBlock)  (None, 112, 112, 32)     3456      
...
```

VS:

```
input_1 (InputLayer)           [(None, 224, 224, 3  0           []                               
                                )]                                                                
                                                                                                  
 rescaling (Rescaling)          (None, 224, 224, 3)  0           ['input_1[0][0]']                
                                                                                                  
 stem_conv (Conv2D)             (None, 112, 112, 32  864         ['rescaling[0][0]']              
                                )                                                                 
                                                                                                  
 stem_bn (BatchNormalization)   (None, 112, 112, 32  128         ['stem_conv[0][0]']              
                                )                                                                 
                                                                                                  
 stem_activation (Activation)   (None, 112, 112, 32  0           ['stem_bn[0][0]']                
                                )                                                                 
# NOW PART OF FusedMBConvBlock                                      
 block1a_project_conv (Conv2D)  (None, 112, 112, 16  4608        ['stem_activation[0][0]']        
                                )                                                                 
                                                                                                  
 block1a_project_bn (BatchNorma  (None, 112, 112, 16  64         ['block1a_project_conv[0][0]']   
 lization)                      )                                                                 
                                                                                                  
 block1a_project_activation (Ac  (None, 112, 112, 16  0          ['block1a_project_bn[0][0]']    
```

@LukeWood @tanzhenyu @ianstenbit @bhack

### CODE REVIEW

**Review by tanzhenyu — COMMENTED** [review]
Thanks for the PR!
Now that they will become public API, please add:
1) unit test files, regular case + corner case
2) serialization test under serialization_test.py
3) better documentation for each layer, i.e., what's the paper reference, what does it do, what inputs / outputs shape does it require, what is the python example usage, etc

**Comment by DavidLandup0** [comment]
Deal! Wanted to check whether this was an okay change to begin with :)
Docs and tests coming up tomorrow

**Comment by tanzhenyu** [comment]
We can still easily remap some of the layers weights when the total number of layer weights changes, I guess?

**Comment by DavidLandup0** [comment]
Yeah, I think that's the best approach. No need to re-train since it's identical, just wrapped. Has a nicer summary call now too :)

**Comment by DavidLandup0** [comment]
Should I do the remapping or should someone here do it instead, since you can directly upload the new weights and verify on the same eval set as before?
I'd have to test on V2 and open a PR which unnecessarily slows the process down

**Comment by ianstenbit** [comment]
> Should I do the remapping or should someone here do it instead, since you can directly upload the new weights and verify on the same eval set as before? I'd have to test on V2 and open a PR which unnecessarily slows the process down

I will re-map the weights today and upload the new ones to GCS

File: keras_cv/layers/fusedmbconv.py [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+        self.se_ratio = se_ratio
+        self.bn_momentum = bn_momentum
+        self.activation = activation
+        self.survival_probability = survival_probability
+        self.filters = self.input_filters * self.expand_ratio
+        self.filters_se = max(1, self.filters)
+
+        self.conv1 = layers.Conv2D(
+            filters=self.filters,
+            kernel_size=1,
```
Comment (ianstenbit) [inline keras_cv/layers/fusedmbconv.py:64]: Looks like this was also `kernel_size` in the original impl

File: keras_cv/layers/mbconv.py [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+        self.output_filters = output_filters
+        self.expand_ratio = expand_ratio
+        self.kernel_size = kernel_size
+        self.strides = strides
+        self.se_ratio = se_ratio
+        self.bn_momentum = bn_momentum
+        self.activation = activation
+        self.survival_probability = survival_probability
+        self.filters = self.input_filters * self.expand_ratio
+        self.filters_se = max(1, self.filters)
```
Comment (ianstenbit) [inline keras_cv/layers/mbconv.py:59]: Should be `max(1, int(input_filters * se_ratio)`

File: keras_cv/layers/fusedmbconv.py [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+        self.output_filters = output_filters
+        self.expand_ratio = expand_ratio
+        self.kernel_size = kernel_size
+        self.strides = strides
+        self.se_ratio = se_ratio
+        self.bn_momentum = bn_momentum
+        self.activation = activation
+        self.survival_probability = survival_probability
+        self.filters = self.input_filters * self.expand_ratio
+        self.filters_se = max(1, self.filters)
```
Comment (ianstenbit) [inline keras_cv/layers/fusedmbconv.py:60]: Should be `max(1, int(input_filters * se_ratio)`

File: keras_cv/layers/fusedmbconv.py [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+
+        self.conv1 = layers.Conv2D(
+            filters=self.filters,
+            kernel_size=1,
+            strides=1,
```
Comment (ianstenbit) [inline keras_cv/layers/fusedmbconv.py:65]: should be `strides=strides`

**Comment by ianstenbit** [comment]
I've re-mapped the weights to the new implementation (with changes I made to your PR based on my comments here). You can see my updates version of your new ConvBlock+FusedMBConvBlocks here: https://github.com/ianstenbit/keras-cv/tree/mbconv

Feel free to copy those changes in. The only change I commented here and didn't update in my branch is the unused depthwise conv filter in the Fused block.

My script to convert the weights can be found [here](), and I've uploaded the weights to GCS. Once this PR is merged, I will update our pre-trained weights config to point to the converted weights.

**Comment by ianstenbit** [comment]
PR to update our pretrained weights after this is merged: #1151

**Comment by DavidLandup0** [comment]
Thanks for the changes and porting Ian!
My bad for not catching them myself. I didn't want to fiddle with the arguments since they come from the current EffNetV2s, so I a priori assumed that they're okay.

I'll copy the changes and attribute them to you and update this PR. Excited to see these disentangled, making EffNets more modular! We should do a sweep like this from time to time to reasses whether architectures are modular enough. :)

File: keras_cv/layers/mbconv.py [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+        else:
+            x = inputs
+
+        # Depthwise conv
+        x = self.depthwise(x)
+        x = self.bn2(x)
+        x = self.act(x)
+
+        # Squeeze and excite
+        if 0 < self.se_ratio <= 1:
```
Comment (DavidLandup0) [inline keras_cv/layers/mbconv.py:143]: Miss here, copied from the original implementation: https://github.com/keras-team/keras-cv/blob/5ea90682ffc2a5e28262d5e2261e8dd9f6bedeb2/keras_cv/models/efficientnet_v2.py#L637

The output phase is conditional, based on the `se_ratio`, which is totally off. Should've been unindented

File: keras_cv/layers/fusedmbconv.py [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+        )
+        self.bn1 = layers.BatchNormalization(
+            axis=BN_AXIS,
+            momentum=self.bn_momentum,
+            name=self.name + "expand_bn",
+        )
+        self.act = layers.Activation(
+            self.activation, name=self.name + "expand_activation"
+        )
+        self.depthwise = layers.DepthwiseConv2D(
```
[the remark this answers is not in the document]
Comment (DavidLandup0) [reply keras_cv/layers/fusedmbconv.py]: Removed

**Comment by DavidLandup0** [comment]
Changed:

- Docs, example usage, links to relevant papers
- Fixed conditional output from the original EffNetV2 implementation here
- Added test cases that would've failed for the original bug
- Addd serialization tests

Thanks for the super fast response @ianstenbit, and for catching the bugs :)

File: keras_cv/layers/fusedmbconv_test.py [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+    def test_different_input_output_shapes(self):
+        inputs = tf.random.normal(shape=(1, 64, 64, 32), dtype=tf.float32)
+        layer = FusedMBConvBlock(input_filters=32, output_filters=48)
+
+        output = layer(inputs)
+        self.assertEquals(output.shape, [1, 64, 64, 48])
+        self.assertLen(output, 1)
+        self.assertTrue(isinstance(output, tf.Tensor))
+
+    def test_se(self):
```
Comment (tanzhenyu) [inline keras_cv/layers/fusedmbconv_test.py:38]: nit: better test naming

**Review by tanzhenyu — APPROVED** [review]
Overall LGTM. Please wait for Ian's comment

File: keras_cv/layers/fusedmbconv_test.py (hunk 7 above)
Comment (DavidLandup0) [reply keras_cv/layers/fusedmbconv_test.py]: updated

**Comment by DavidLandup0** [comment]
This has nothing to do with these changes, right? 🤔 

```
keras_cv/layers/preprocessing/auto_contrast_test.py FFFF.s               [ 39%]
keras_cv/layers/preprocessing/base_image_augmentation_layer_test.py FFFF [ 40%]
F.FFF.FFFs                                                               [ 41%]
keras_cv/layers/preprocessing/channel_shuffle_test.py ......s            [ 41%]
keras_cv/layers/preprocessing/cut_mix_test.py FFFFFF.sF                  [ 42%]
```

**Comment by sebastian-sz** [comment]
Thanks for catching this. This is clearly a mistake on my end during the
rewrite.

Upon inspection it seems this error never happens as MBConvBlock luckily
always has se_ratio equal to 0.25 which executes the final conv layer.

This should be absolutely fixed but it should not affect converted weights
nor existing models (here and in tf.keras.applications).



pt., 16 gru 2022, 04:30 użytkownik David Landup ***@***.***>
napisał:

File: keras_cv/layers/mbconv.py (hunk 5 above)
Comment (sebastian-sz) [reply keras_cv/layers/mbconv.py:185]: Thanks for catching this. This is clearly a mistake on my end during the
rewrite.

Upon inspection it seems this error never happens as MBConvBlock luckily
always has se_ratio equal to 0.25 which executes the final conv layer.

This should be absolutely fixed but it should not affect converted weights
nor existing models (here and in tf.keras.applications)

Comment (DavidLandup0) [reply keras_cv/layers/mbconv.py:185]: Thanks for weighing in! I saw some `se_ratio`s being 0.0, but didn't check the conv type. All MBConv blocks had `se_ratio`>0! The FusedMBConv blocks didn't have this issue :)

This shouldn't affect the weight mapping or performance then. The PR should be ready for a merge if everything else looks okay @tanzhenyu

**Comment by ianstenbit** [comment]
> This has nothing to do with these changes, right? 🤔
> 
> ```
> keras_cv/layers/preprocessing/auto_contrast_test.py FFFF.s               [ 39%]
> keras_cv/layers/preprocessing/base_image_augmentation_layer_test.py FFFF [ 40%]
> F.FFF.FFFs                                                               [ 41%]
> keras_cv/layers/preprocessing/channel_shuffle_test.py ......s            [ 41%]
> keras_cv/layers/preprocessing/cut_mix_test.py FFFFFF.sF                  [ 42%]
> ```

Looks unrelated -- can you merge from keras-team/master into your branch and re-push here?

**Comment by ianstenbit** [comment]
> Thanks for the changes and porting Ian! My bad for not catching them myself. I didn't want to fiddle with the arguments since they come from the current EffNetV2s, so I a priori assumed that they're okay.
> 
> I'll copy the changes and attribute them to you and update this PR. Excited to see these disentangled, making EffNets more modular! We should do a sweep like this from time to time to reasses whether architectures are modular enough. :)

No problem -- it was actually a pretty fun challenge to get these migrated over.

I've re-run the weight migration and a/b testing using your latest code, and it all looks good. Thank you!

**Review by ianstenbit — APPROVED** [review]
LGTM -- thank you!

**Comment by DavidLandup0** [comment]
> Looks unrelated -- can you merge from keras-team/master into your branch and re-push here?

Just did - but it's showing the same failures on `CutMix` and some other preprocessing layers. 🤔 
It's passing locally:
image

> No problem -- it was actually a pretty fun challenge to get these migrated over.
I've re-run the weight migration and a/b testing using your latest code, and it all looks good. Thank you!

Glad to hear it was fun and happy to have been able to contribute :)
MaxViT blocks with @ayulockin coming in soon!

**Comment by tanzhenyu** [comment]
> Details

@ianstenbit why is this related to keras?

**Comment by DavidLandup0** [comment]
@tanzhenyu I'm pretty sure he meant KCV's master branch, since there were recent changes to KPL's. KCV's master branch is also 'keras-team:master'

**Comment by DavidLandup0** [comment]
All tests pass locally and the PR doesn't change anything regarding CutMix and BaseAugmenter. Could you test this locally  as well @tanzhenyu?

**Comment by tanzhenyu** [comment]
> All tests pass locally and the PR doesn't change anything regarding CutMix and BaseAugmenter. Could you test this locally as well @tanzhenyu?

I don't think this PR breaks any CutMix related test, is there any way you can re-sync and run it again?

**Comment by DavidLandup0** [comment]
Merged the latest changes, tried a new PR, ran tests locally again - the CI is still showing failed tests on some preprocessing layers. No clue why 🤔

**Comment by tanzhenyu** [comment]
> Merged the latest changes, tried a new PR, ran tests locally again - the CI is still showing failed tests on some preprocessing layers. No clue why 🤔

Any luck with the new PR 1168?

**Comment by bhack** [comment]
> Merged the latest changes, tried a new PR, ran tests locally again - the CI is still showing failed tests on some preprocessing layers. No clue why thinking

@DavidLandup0  You can reproduce the same issue/failures locally with the full test sequence:

```bash
pytest keras_cv/ --ignore keras_cv/models --ignore keras_cv/datasets/waymo

```

**Comment by DavidLandup0** [comment]
> @DavidLandup0 You can reproduce the same issue/failures locally with the full test sequence:
> 
> ```shell
> pytest keras_cv/ --ignore keras_cv/models --ignore keras_cv/datasets/waymo
> ```

Hmm, the individual test runs: 

```
keras_cv/layers/preprocessing/base_image_augmentation_layer_test.py .............s
```

But it fails when done with the rest of the tests. I'll create a fresh branch and make the changes there...

**Comment by DavidLandup0** [comment]
Found it. When calling:

```
output = layer(inputs)
```

It *has to* be:

```
output = layer(inputs, training=True)
```

Otherwise, the tests pass individually, but don't pass in the suite, presumably because some state might be reused between unit tests. When run for a single file, the state is fine because it's localized and can't affect other files. When run in the suite, the change in the state triggers a failure in other tests.

If this is true, we might want to add a before clause in the unit tests to refresh whatever caused this? @LukeWood @bhack @tanzhenyu

**Comment by tanzhenyu** [comment]
> Found it. When calling:
> 
> ```
> output = layer(inputs)
> ```
> 
> It _has to_ be:
> 
> ```
> output = layer(inputs, training=True)
> ```
> 
> Otherwise, the tests pass individually, but don't pass in the suite, presumably because some state might be reused between unit tests. When run for a single file, the state is fine because it's localized and can't affect other files. When run in the suite, the change in the state triggers a failure in other tests.
> 
> If this is true, we might want to add a before clause in the unit tests to refresh whatever caused this? @LukeWood @bhack @tanzhenyu
> 
> P.S. Might be worth adding this in the contributor guidelines if it's unfixable for whatever reason, for contributors to know upfront, because it's a very silent failure across 150 tests caused by _one_ keyword in a fully separate test. 🤔

I suspect this might be caused by a bug. The preprocessing layer uses `training=True` by default, whereas the regular layers uses `training=None` by default.
Other preprocessing layers don't use this argument in test: https://github.com/keras-team/keras-cv/blob/master/keras_cv/layers/preprocessing/cut_mix_test.py
Other regular layers don't use this argument in test: https://github.com/keras-team/keras-cv/blob/master/keras_cv/layers/object_detection/non_max_suppression_test.py

**Comment by DavidLandup0** [comment]
When was training set to True by default for the KPLs?
A few months back, I remember having an issue where they'd be True for the first execution, but False for all subsequent ones, so when visualizing training batches, the augmentations would be off during training, and I had to manually set training as True during batch visualization to avoid this.

It was tricky to reproduce, and only happened some times on Google Colab (but not locally) so I never opened an issue, and it eventually stopped. Might be tied to this somehow?

**Comment by tanzhenyu** [comment]
> When was training set to True by default for the KPLs? A few months back, I remember having an issue where they'd be True for the first execution, but False for all subsequent ones, so when visualizing training batches, the augmentations would be off during training, and I had to manually set training as True during batch visualization to avoid this.
> 
> It was tricky to reproduce, and only happened some times on Google Colab (but not locally) so I never opened an issue, and it eventually stopped. Might be tied to this somehow?

https://github.com/keras-team/keras-cv/blob/master/keras_cv/layers/preprocessing/base_image_augmentation_layer.py#L362

**Comment by tanzhenyu** [comment]
> When was training set to True by default for the KPLs? A few months back, I remember having an issue where they'd be True for the first execution, but False for all subsequent ones, so when visualizing training batches, the augmentations would be off during training, and I had to manually set training as True during batch visualization to avoid this.
> 
> It was tricky to reproduce, and only happened some times on Google Colab (but not locally) so I never opened an issue, and it eventually stopped. Might be tied to this somehow?

Can you try something like this?
https://github.com/keras-team/keras-cv/blob/master/keras_cv/models/object_detection/retina_net/retina_net_test.py#L27

**Comment by bhack** [comment]
I think It is better to introduce a pytest fixture 

https://github.com/BenWhetton/keras-surgeon/blob/master/tests/test_surgeon.py#L31-L38

**Comment by DavidLandup0** [comment]
Added fixture to clean the session - good call 👍 
@bhack @tanzhenyu

**Comment by tanzhenyu** [comment]
> 

Consider adding this cmd to the contribution guideline :-)

**Comment by bhack** [comment]
@tanzhenyu Can we find a more general placement? Is this still used in keras https://github.com/keras-team/keras/pull/11170?

**Comment by bhack** [comment]
Do we have switched over `teardown` func https://github.com/keras-team/keras/blob/b8c48898a3f316dda43231a39c21fc4a61aed3b1/keras/layers/preprocessing/integer_lookup_test.py#L614-L617 ?

**Comment by tanzhenyu** [comment]
The failing gcp test is due to custom ops build issue. Manually merging

**Comment by tanzhenyu** [comment]
> @tanzhenyu Can we find a more general placement? Is this still used in keras [keras-team/keras#11170](https://github.com/keras-team/keras/pull/11170)?

Looks like a good thing to add to our codebase
